### DISCUSSION TITLE
How to use bigger font?

### DISCUSSION THREAD

**Discussion by jd3096-mpy** [shorepine/tulipcc#49] [General]
My plan is to use 5.0 inch 800x480 screen ,but now font is too small (8x12).
I find out the font file smallfont.c and the font setting #define FONT_HEIGHT 12  #define FONT_WIDTH 8.
So can I changed other bigger font?Maybe 12x24 font
The first line in smallfont.c is 
const unsigned char font_8x12_r[256][12]
Where can I found font files in this format?Or can you provided a 12x24 font file or font convert tools？ Thanks.

**Comment by bwhitman** [comment]
That font has been in Tulip for a while. I think I got it from something like https://int10h.org/oldschool-pc-fonts/ where you'll also want to use https://lvgl.io/tools/fontconverter to convert to the .h file. The _r at the end means I've reversed the bits of each character for faster drawing on Tulip (we draw scanline at a time.) Adding a new TFB font to Tulip sounds like a fun challenge, do let us know how it goes!

**Reply by jd3096-mpy** [reply]
> That font has been in Tulip for a while. I think I got it from something like https://int10h.org/oldschool-pc-fonts/ where you'll also want to use https://lvgl.io/tools/fontconverter to convert to the .h file. The _r at the end means I've reversed the bits of each character for faster drawing on Tulip (we draw scanline at a time.) Adding a new TFB font to Tulip sounds like a fun challenge, do let us know how it goes!

I tried others fonts like 12x20 12x16 but failed,So I tried other screen 4.3inch 480 272 ,Now fonts seems good,but screen like this:
![5ebaf903c9197dee356b66288fd9580](https://user-images.githubusercontent.com/69076780/229403675-39c77cc9-fee0-4838-92be-db3e242bc74e.jpg)
There are many mistakes in 480x272 resolution? I tried 800x480 resolution settings on 480 272 screen, it can work normally.

**Reply by HalfSweet** [reply]
Yes, I also tried a screen with 480*272 resolution and the same image misalignment appeared
It seems we can use more or less pixels to avoid the appearance, for example 264 or 278, at least a multiple of 12 seems to work fine without problems
But if more pixels are used to fill, the whole bottom of the screen will be displayed incompletely; if fewer pixels are used, then a certain area of the display will be lost.
This seems to be a font rendering problem, is it possible to get a solution?
![图片](https://user-images.githubusercontent.com/60973476/230136976-1841b125-0d8e-4c73-b956-f7ed18950246.png)
@bwhitman

**Comment by bwhitman** [comment]
@HalfSweet I'm not sure if I understand the problem. Is that pic a 480x272 screen? The last line will not draw properly because 272 is not a multiple of 12. What sort of solution do you propose? I'm unclear what "more pixels are used to fill" means in this context. You could change the TFB font (in `smallfont.c` to a 8x8 font, and change `BOUNCE_BUFFER_SIZE_PX` accordingly, and everything should fit fine in a 480x272 at 60x34.

**Reply by HalfSweet** [reply]
Yes, your solution is good, I tried to set the font width and height to 8, the screen works fine no more misalignment, but what confuses me is how should I go about making an array of fonts like `font_8x12_r[256][12]`? It doesn't look like a pure ASCII-only font, it also encodes other content.

Also I see that you can select more fonts for display in the updated image in the api manual, which looks really cool! But I don't see how I can call it and how I should make my own fonts for display.

Thank you very much for your help!

**Reply by bwhitman** [reply]
The fonts we added last week are for drawing to the BG, not the TFB.

The TFB is "special" because it is drawn in real time on top of the BG as the screen refreshes (at 30Hz by default.) This lets fast text update for things like the REPL and text editors. We store only the ASCII codes in fast ram and the screen refresh function does the work of drawing the text characters as the lines are drawn to the screen. (This is quite like the IBM PC and similar 80s machines.) The font for the TFB is not like the rest as it has to be fixed width and has to be drawn very quickly. I didn't think of ways to easily change this font because it's tied directly to the screen refresh code. It's definitely possible! But would take some effort to set up.

**Reply by HalfSweet** [reply]
hmmm...this sounds interesting, but do you know what each of the 256 characters corresponds to? I guess I could start by writing a script to adapt the 8*8 font

**Reply by bwhitman** [reply]
It's an ASCII table https://www.ascii-code.com
